### ISSUE TITLE
sharing Alpaca-LoRA chatbot (Gradio app)

### ISSUE THREAD

**Issue by deep-diver** [tloen/alpaca-lora#87]
I hope you guys find it useful. Here is the repo
: https://github.com/deep-diver/Alpaca-LoRA-Serve

and I am currently running in on my VM. Try it while it is live
: https://notebooksf.jarvislabs.ai/43j3x9FSS8Tg0sqvMlDgKPo9vsoSTTKRsX4RIdC3tNd6qeQ6ktlA0tyWRAR3fe_l/

**Comment by DanielWe2** [comment]
Thanks. I have played around to test its German capabilities. It's better than the 7b version that is great. Not sure if that is caused by the bigger dataset with that Llama 30b was grained with compared to the two smaller versions or if the bigger model is the reason.

In my testing of multiple messages it only made one grammar error. Not perfect but it way better than 7b which didn't sound like a true native speaker of German.

Is that an 8 bit or 4bit version of the model?

And how does your code compare to the "text generation web ui" repo?

**Comment by espeon** [comment]
This is really cool! Messed around with it earlier and was just wondering - what gpu(s) is this running on?

**Comment by deep-diver** [comment]
8bit one running on A5000

**Comment by deep-diver** [comment]
Not sure about web ui project. Just I have built this for real serving need. Batch request processing and some handy buttons like summarize and continue to overcome low resource GPU problem.

**Comment by DanielWe2** [comment]
Ok, is it possible that there is a commit missing in the app. Py? Can not find the place where it adds the summarize and continue buttons.

**Comment by deep-diver** [comment]
Dang I missed to upload it. Will share when I commit

**Comment by deep-diver** [comment]
@BEpresent that is something I don't know of. we need some experimentations.
@DanielWe2 the code has been updated

**Comment by deep-diver** [comment]
@makovez 
the same dataset (actually the one cleaned-up in this repository).

**Comment by deep-diver** [comment]
@makovez 

you can actually use the notebook in my repository. here is the [link](https://github.com/deep-diver/Alpaca-LoRA-Serve/blob/main/notebooks/alpaca_lora_in_colab.ipynb). just remember to change the checkpoint right
- Collab pro w/ premium GPU option would work with 30B
- otherwise, choose 13B or 7B

**Comment by deep-diver** [comment]
@makovez This is just a playground. In order to reflect more reality, it should be changed a lot. 

My basic idea is to let users to choose when to summarize or not. Since this is running on a single GPU + Int8, the inference speed is not so good. Hence, it wouldn't be possible to let people just wait for a very long time until all the tokens are generated. Also, it gets even worse when there are lots of traffic.

For this reason, I wanted to build a bit better UI for people to interact with. I set `timeout=30`, so the reponse might not be completed, but I wanted to give some handy option like `continue` and `summarize` to go on.

**Comment by deep-diver** [comment]
there should be lots of tweaks on pre/post processings. For instance, when the response is written in Markdown format, it wouldn't be rendered properly. It should be converted in HTML format. Also, the HTML converted text should be re-converted back to the original text since lots of HTML tags wouldn't be helpful for LLaMA to understand the context.

**Comment by deep-diver** [comment]
Yeap. at some point ChatGPT should be doing this since there is limit max input size. it depends how accurately remembers the details of the previous conversations.

**Comment by DanielWe2** [comment]
> But do you think this is what chatgpt does? Hard code previous Q/A in new input?
> 
> Because otherwise how can it remember context, for each different chat of user?

I don't think ChatGPT does remember previous sessions at all. (at least the last time I used it). Bing doesn't.

If you want some kind of persitence memory you would most likely connect it to some kind of database.

LangChains does provide infrastrucute for that. For different kind of memorys for ChatBots.  I think it would be good idea to base ChatBot on top of that, it already solved a lot of thoses problems and also has tools to add differernt serivce (online search or database search for example) to a chat bot.

Take a look at https://langchain.readthedocs.io/en/latest/modules/memory.html  and the sub chapters on the left for different type of local or long term memory.

**Comment by deep-diver** [comment]
Sounds great. I think it is pretty much doable. Just need to figure out how to make this happen fast enough. As you see, it takes about 20-30 seconds to generate a handful of text. 

If I input many many texts into it, it gets even slower. So, I guess I need to find a way to speed up before applying memory structure.

**Comment by makovez** [comment]
> Sounds great. I think it is pretty much doable. Just need to figure out how to make this happen fast enough. As you see, it takes about 20-30 seconds to generate a handful of text. 
> 
> If I input many many texts into it, it gets even slower. So, I guess I need to find a way to speed up before applying memory structure. 

How do you plan to speed it up?

**Comment by deep-diver** [comment]
Need to explore every possible combinations of different hyperparameters + float16 instead of 8int

**Comment by BEpresent** [comment]
> @BEpresent that is something I don't know of. we need some experimentations. @DanielWe2 the code has been updated

I get something like this, when adding a context. I'm not sure if I made any obvious mistake?

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/7645717/226458411-5db4f97c-bdae-4885-8734-171345b9eb69.png">

**Comment by makovez** [comment]
> > @BEpresent that is something I don't know of. we need some experimentations. @DanielWe2 the code has been updated
> 
> I get something like this, when adding a context. I'm not sure if I made any obvious mistake?
> 
> <img width="1102" alt="image" src="https://user-images.githubusercontent.com/7645717/226458411-5db4f97c-bdae-4885-8734-171345b9eb69.png">
> 

Actually context doesn't seem to work. ![Screenshot_2023-03-20-21-42-57-94_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/21694707/226461396-f0b5d837-af63-4a25-a9c6-174bd7b61a2c.jpg)

But doing this works.

**Comment by BEpresent** [comment]
Right, it seems to ignore the initial context, but seems to pick it up from the chat if mentioned

**Comment by deep-diver** [comment]
Thanks for the feedback! Need to check how to craft better prompt template

**Comment by deep-diver** [comment]
Updated! Check out the new issue

**Comment by wierceniestudnimazowieckie** [comment]
Thanks!

**Comment by Meghansh-Maurya** [comment]
> > > @BEpresent that is something I don't know of. we need some experimentations. @DanielWe2 the code has been updated
> > 
> > 
> > I get something like this, when adding a context. I'm not sure if I made any obvious mistake?
> > <img alt="image" width="1102" src="https://user-images.githubusercontent.com/7645717/226458411-5db4f97c-bdae-4885-8734-171345b9eb69.png">
> 
> Actually context doesn't seem to work. ![Screenshot_2023-03-20-21-42-57-94_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/21694707/226461396-f0b5d837-af63-4a25-a9c6-174bd7b61a2c.jpg)
> 
> But doing this works.
> 
> Ps. Not sure where that John Smith comes from but ok🤣

Did you solve this issue? I have trained it on some other name but I am getting the same name(John Smith)
